# STUDENT NAME : Kunga Nyima Gurung
# COURSEWORK-1 : DECISION TRESS USING ID3 ALGORITHM
# WLV ID : 2050453 


# ID3 ALGORITHM

## FORMULA

![](Entropy_formula.png)

## FORMULA

![](Information_Gain.png)

# Implement Decision Tree(ID3 Algorithm) from scratch

In [1]:
from pandas.errors import InvalidIndexError
#Importing libraries
import numpy as np
import pandas as pd

#ignore warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
#Reading the dataset
dataset = pd.read_csv("play_tennis.csv")
dataset.head(15)

outlook  temp humidity    wind play
0      Sunny   Hot     High    Weak   No
1      Sunny   Hot     High  Strong   No
2   Overcast   Hot     High    Weak  Yes
3       Rain  Mild     High    Weak  Yes
4       Rain  Cool   Normal    Weak  Yes
5       Rain  Cool   Normal  Strong   No
6   Overcast  Cool   Normal  Strong  Yes
7      Sunny  Mild     High    Weak   No
8      Sunny  Cool   Normal    Weak  Yes
9       Rain  Mild   Normal    Weak  Yes
10     Sunny  Mild   Normal  Strong  Yes
11  Overcast  Mild     High  Strong  Yes
12  Overcast   Hot   Normal    Weak  Yes
13      Rain  Mild     High  Strong   No

# THE FIRST ATTRIBUTE "Outlook"

In [3]:
#unique Values in the attributr outlook
outlook_values = dataset["outlook"].unique()
print(outlook_values)

['Sunny' 'Overcast' 'Rain']


In [4]:
values = dataset.groupby('outlook')['outlook'].count()
print (values)

outlook
Overcast    4
Rain        5
Sunny       5
Name: outlook, dtype: int64


Here, in the atttribute "outlook" the possible values are: "Sunny", "Overcast" and "Rain".

In [5]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 9/14
class_no = 5/14
# calculate entropy
entropy = -(class_yes * log2(class_yes) + class_no * log2(class_no))
# print the result
print('entropy: %.2f' % entropy)

entropy: 0.94


## SUNNY_ENTROPY

In [6]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 2/5
class_no = 3/5
# calculate entropy
Sunny_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Sunny_entropy)

entropy: 0.971


## RAIN_ENTROPY

In [7]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 3/5
class_no = 2/5
# calculate entropy
Rain_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Rain_entropy)

entropy: 0.971


## OUTLOOK_ENTROPY

In [8]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 4/4
class_no = 0/4
# calculate entropy
Outlook_entropy = (-class_yes) * log2(class_yes) - 0
# print the result
print('entropy: %.3f' % Outlook_entropy)

entropy: -0.000


# Calculating information gain for feature "outlook"

In [9]:
# calculate the information gain
p_sunny=5/14
p_Rain=5/14
p_Outlook=4/14
Outlook_Average_entropy = (p_sunny * Sunny_entropy) + (p_Rain * Rain_entropy) + (p_Outlook * Outlook_entropy)
print('Average Entropy : %.2f' % Outlook_Average_entropy)

Average Entropy : 0.69


In [10]:
Outlook_Information_Gain = entropy - Outlook_Average_entropy
print('Information Gain : %.3f' % Outlook_Information_Gain)

Information Gain : 0.247


# THE SECOND ATTRIBUTE "temp"

In [11]:
outlook_values = dataset["temp"].unique()
print(outlook_values)

['Hot' 'Mild' 'Cool']


In [12]:
values = dataset.groupby('temp')['temp'].count()
print (values)

temp
Cool    4
Hot     4
Mild    6
Name: temp, dtype: int64


## COOL_ENTROPY

In [13]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 3/4
class_no = 1/4
# calculate entropy
Cool_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Cool_entropy)

entropy: 0.811


## HOT_ENTROPY

In [14]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 2/4
class_no = 2/4
# calculate entropy
Hot_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Hot_entropy)

entropy: 1.000


## MILD_ENTROPY

In [15]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 4/6
class_no = 2/6
# calculate entropy
Mild_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Mild_entropy)

entropy: 0.918


# Calculating information gain for feature "temp"

In [16]:
# calculate the information gain
p_Cool=4/14
p_Hot=4/14
p_Mild=6/14
Temp_Average_entropy = (p_Cool * Cool_entropy) + (p_Hot * Hot_entropy) + (p_Mild * Mild_entropy)
print('Average Entropy : %.2f' % Temp_Average_entropy)

Average Entropy : 0.91


In [17]:
Temp_Information_Gain = entropy - Temp_Average_entropy
print('Information Gain : %.3f' % Temp_Information_Gain)

Information Gain : 0.029


# THE THIRD ATTRIBUTE "humidity"

In [18]:
outlook_values = dataset["humidity"].unique()
print(outlook_values)

['High' 'Normal']


In [19]:
values = dataset.groupby('humidity')['humidity'].count()
print (values)

humidity
High      7
Normal    7
Name: humidity, dtype: int64


## HIGH_ENTROPY

In [20]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 3/7
class_no = 4/7
# calculate entropy
High_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % High_entropy)

entropy: 0.985


## NORMAL_ENTROPY 

In [21]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 6/7
class_no = 1/7
# calculate entropy
Normal_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Normal_entropy)

entropy: 0.592


# Calculating information gain for feature "humidity"

In [22]:
# calculate the information gain
p_High=7/14
p_Normal=7/14
Humidity_Average_entropy = (p_High * High_entropy) + (p_Normal * Normal_entropy) 
print('Average Entropy : %.3f' % Humidity_Average_entropy)

Average Entropy : 0.788


In [23]:
Humidity_Information_Gain = entropy - Humidity_Average_entropy
print('Information Gain : %.3f' % Humidity_Information_Gain)

Information Gain : 0.152


# THE FOURTH ATTRIBUTE "wind"

In [24]:
outlook_values = dataset["wind"].unique()
print(outlook_values)

['Weak' 'Strong']


In [25]:
values = dataset.groupby('wind')['wind'].count()
print (values)

wind
Strong    6
Weak      8
Name: wind, dtype: int64


## STRONG_ENTROPY

In [26]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 3/6
class_no = 3/6
# calculate entropy
Strong_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Strong_entropy)

entropy: 1.000


## WEAK-ENTROPY

In [27]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 6/8
class_no = 2/8
# calculate entropy
Weak_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Weak_entropy)

entropy: 0.811


# Calculating information gain for feature "wind"

In [28]:
# calculate the information gain
p_Strong=6/14
p_Weak=8/14
Wind_Average_entropy = (p_Strong * Strong_entropy) + (p_Weak * Weak_entropy) 
print('Average Entropy : %.3f' % Wind_Average_entropy)

Average Entropy : 0.892


In [29]:
Wind_Information_Gain = entropy - Wind_Average_entropy
print('Information Gain : %.3f' % Wind_Information_Gain)

Information Gain : 0.048


# Finding the most informative Attributes

In [30]:
Outlook_Gain = Outlook_Information_Gain
Temp_Gain =  Temp_Information_Gain
Humidity_Gain = Humidity_Information_Gain
Wind_Gain = Wind_Information_Gain


print('Gain(S, Outlook) : %.3f' % Outlook_Gain)
print('Gain(S, Temp) : %.3f' % Temp_Gain)
print('Gain(S, Humidity) : %.3f' % Humidity_Gain)
print('Gain(S, Wind) : %.3f' % Wind_Gain)

Gain(S, Outlook) : 0.247
Gain(S, Temp) : 0.029
Gain(S, Humidity) : 0.152
Gain(S, Wind) : 0.048


As we can see that the attribute "Outlook" has the hightst gain of 0.247, we will consider "outlook as a root node in this case.

![](tree_dgm.png)

When we see the label of sunny and wind, it is in terms of positive and negative label. So we cannot write yes or no. Whereas, in case of overcast all the label are positive so we write yes.

In [31]:
dataset = pd.read_csv("play_tennis.csv")
dataset.head(15)

outlook  temp humidity    wind play
0      Sunny   Hot     High    Weak   No
1      Sunny   Hot     High  Strong   No
2   Overcast   Hot     High    Weak  Yes
3       Rain  Mild     High    Weak  Yes
4       Rain  Cool   Normal    Weak  Yes
5       Rain  Cool   Normal  Strong   No
6   Overcast  Cool   Normal  Strong  Yes
7      Sunny  Mild     High    Weak   No
8      Sunny  Cool   Normal    Weak  Yes
9       Rain  Mild   Normal    Weak  Yes
10     Sunny  Mild   Normal  Strong  Yes
11  Overcast  Mild     High  Strong  Yes
12  Overcast   Hot   Normal    Weak  Yes
13      Rain  Mild     High  Strong   No

# SUNNY

In [32]:
Sunny_dataset = dataset.loc[[0, 1, 7,8,10], :]
Sunny_dataset. drop('outlook', axis=1, inplace=True)
Sunny_dataset.head()

temp humidity    wind play
0    Hot     High    Weak   No
1    Hot     High  Strong   No
7   Mild     High    Weak   No
8   Cool   Normal    Weak  Yes
10  Mild   Normal  Strong  Yes

In [33]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 2/5
class_no = 3/5
# calculate entropy
Sunny_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Sunny_entropy)

entropy: 0.971


## THE FIRST ATTRIBUTE "temp"

In [34]:
temp_values = Sunny_dataset["temp"].unique()
print(temp_values)

['Hot' 'Mild' 'Cool']


In [35]:
values = Sunny_dataset.groupby('temp')['temp'].count()
print (values)

temp
Cool    1
Hot     2
Mild    2
Name: temp, dtype: int64


## COOL_ENTROPY

In [36]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 1/1
class_no = 0/1
# calculate entropy
Cool_entropy = -(class_yes * log2(class_yes) - class_no * 0)
# print the result
print('entropy: %.3f' % Cool_entropy)

entropy: -0.000


## HOT_ENTROPY

In [37]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 0/2
class_no = 2/2
# calculate entropy
Hot_entropy = (-class_yes * 0 - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Hot_entropy)

entropy: -0.000


## MILD_ENTROPY 

In [38]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 1/2
class_no = 1/2
# calculate entropy
Mild_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Mild_entropy)

entropy: 1.000


## Calculating information gain for feature "temp"

In [39]:
# calculate the information gain
p_Cool=1/5
p_Hot=2/5
p_Mild=2/5
Average_entropy = (p_Cool * Cool_entropy) + (p_Hot * Hot_entropy)+ (p_Mild * Mild_entropy)  
print('Average Entropy : %.3f' % Average_entropy)

Average Entropy : 0.400


In [40]:
Temp_Info_Gain = Sunny_entropy - Average_entropy
print('Information Gain : %.3f' % Temp_Info_Gain)

Information Gain : 0.571


# THE SECOND ATTRIBUTE "humidity"

In [41]:
humidity_values = Sunny_dataset["humidity"].unique()
print(humidity_values)

['High' 'Normal']


In [42]:
values = Sunny_dataset.groupby('humidity')['humidity'].count()
print (values)

humidity
High      3
Normal    2
Name: humidity, dtype: int64


## HIGH_ENTROPY

In [43]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 0/3
class_no = 3/3
# calculate entropy
High_entropy = -(class_yes * 0 - class_no * 0)
# print the result
print('entropy: %.3f' % High_entropy)

entropy: -0.000


## NORMAL_ENTROPY

In [44]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 2/2
class_no = 0/2
# calculate entropy
Normal_entropy = -(class_yes * log2(class_yes) - class_no * 0)
# print the result
print('entropy: %.3f' % Normal_entropy)

entropy: -0.000


# Calculating information gain for feature "humidity"

In [45]:
# calculate the information gain
p_High=3/5
p_Normal=2/5
Average_entropy = (p_High * High_entropy) + (p_Normal * Normal_entropy)  
print('Average Entropy : %.3f' % Average_entropy)

Average Entropy : -0.000


In [46]:
Humidity_Info_Gain = Sunny_entropy - Average_entropy
print('Information Gain : %.3f' % Humidity_Info_Gain)

Information Gain : 0.971


# THE THIRD ATTRIBUTE "wind"

In [47]:
wind_values = Sunny_dataset["wind"].unique()
print(wind_values)

['Weak' 'Strong']


In [48]:
values = Sunny_dataset.groupby('wind')['wind'].count()
print (values)

wind
Strong    2
Weak      3
Name: wind, dtype: int64


## STRONG_ENTROPY

In [49]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 1/2
class_no = 1/2
# calculate entropy
Strong_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Strong_entropy)

entropy: 1.000


## WEAK_ENTROPY

In [50]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 1/3
class_no = 2/3
# calculate entropy
Weak_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Weak_entropy)

entropy: 0.918


# Calculating information gain for feature "humidity"

In [51]:
# calculate the information gain
p_Strong=2/5
p_Weak=3/5
Average_entropy = (p_Strong * Strong_entropy) + (p_Weak * Weak_entropy)  
print('Average Entropy : %.3f' % Average_entropy)

Average Entropy : 0.951


In [52]:
Wind_Info_Gain = Sunny_entropy - Average_entropy
print('Information Gain : %.3f' % Wind_Info_Gain)

Information Gain : 0.020


# Finding the most informative Attributes

In [53]:
Temp_Gain =  Temp_Info_Gain
Humidity_Gain = Humidity_Info_Gain
Wind_Gain = Wind_Info_Gain


print('Gain(S, Temp) : %.3f' % Temp_Gain)
print('Gain(S, Humidity) : %.3f' % Humidity_Gain)
print('Gain(S, Wind) : %.3f' % Wind_Gain)

Gain(S, Temp) : 0.571
Gain(S, Humidity) : 0.971
Gain(S, Wind) : 0.020


We can see that the label "humidity has the highest information gain of 0.971. Hence we can consider it as a root node.

In [54]:
Sunny_dataset.head()

temp humidity    wind play
0    Hot     High    Weak   No
1    Hot     High  Strong   No
7   Mild     High    Weak   No
8   Cool   Normal    Weak  Yes
10  Mild   Normal  Strong  Yes

From the dataset, we can know that whenever there is high humidity, the label is No and whenever the humidity is normal, the label is yes. Hence the tree can be drawn as:

![](tree_diagram.png)

# RAIN

In [55]:
Rain_dataset = dataset.loc[[3, 4, 5,9,13], :]
Rain_dataset. drop('outlook', axis=1, inplace=True)
Rain_dataset.head()

temp humidity    wind play
3   Mild     High    Weak  Yes
4   Cool   Normal    Weak  Yes
5   Cool   Normal  Strong   No
9   Mild   Normal    Weak  Yes
13  Mild     High  Strong   No

In [56]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 3/5
class_no = 2/5
# calculate entropy
Rain_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Rain_entropy)

entropy: 0.971


## THE FIRST ATTRIBUTE "temp"

In [57]:
temp_values = Rain_dataset["temp"].unique()
print(temp_values)

['Mild' 'Cool']


In [58]:
values = Rain_dataset.groupby('temp')['temp'].count()
print (values)

temp
Cool    2
Mild    3
Name: temp, dtype: int64


## COOL_ENTROPY

In [59]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 1/2
class_no = 1/2
# calculate entropy
Cool_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Cool_entropy)

entropy: 1.000


## MILD_ENTROPY

In [60]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 2/3
class_no = 1/3
# calculate entropy
Mild_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Mild_entropy)

entropy: 0.918


# Calculating information gain for feature "temp"

In [61]:
# calculate the information gain
p_Cool=2/5
p_Mild=3/5
Average_entropy = (p_Cool * Cool_entropy) + (p_Mild * Mild_entropy)  
print('Average Entropy : %.3f' % Average_entropy)

Average Entropy : 0.951


In [62]:
Temp_Info_Gains = Sunny_entropy - Average_entropy
print('Information Gain : %.3f' % Temp_Info_Gains)

Information Gain : 0.020


# THE SECOND ATTRIBUTE "humidity"

In [63]:
humidity_values = Rain_dataset["humidity"].unique()
print(humidity_values)

['High' 'Normal']


In [64]:
values = Rain_dataset.groupby('humidity')['humidity'].count()
print (values)

humidity
High      2
Normal    3
Name: humidity, dtype: int64


# HIGH_ENTROPY

In [65]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 1/2
class_no = 1/2
# calculate entropy
High_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % High_entropy)

entropy: 1.000


# NORMAL_ENTROPY

In [66]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 2/3
class_no = 1/3
# calculate entropy
Normal_entropy = (-class_yes * log2(class_yes) - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Normal_entropy)

entropy: 0.918


# Calculating information gain for feature "temp"

In [67]:
# calculate the information gain
p_Cool=2/5
p_Mild=3/5
Average_entropy = (p_Cool * Cool_entropy) + (p_Mild * Mild_entropy)  
print('Average Entropy : %.3f' % Average_entropy)

Average Entropy : 0.951


In [68]:
Humidity_Info_Gains = Sunny_entropy - Average_entropy
print('Information Gain : %.3f' % Humidity_Info_Gains)

Information Gain : 0.020


# THE THIRD ATTRIBUTE "wind"

In [69]:
humidity_values = Rain_dataset["wind"].unique()
print(humidity_values)

['Weak' 'Strong']


In [70]:
values = Rain_dataset.groupby('wind')['wind'].count()
print (values)

wind
Strong    2
Weak      3
Name: wind, dtype: int64


## STRONG_ENTROPY

In [71]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 0/2
class_no = 2/2
# calculate entropy
Strong_entropy = (-class_yes * 0 - class_no * log2(class_no))
# print the result
print('entropy: %.3f' % Strong_entropy)

entropy: -0.000


## WEAK_ENTROPY

In [72]:
# calculate the entropy for a dataset
from math import log2
# proportion of examples in each class
class_yes = 3/3
class_no = 0/3
# calculate entropy
Weak_entropy = (-class_yes * log2(class_yes) - class_no * 0)
# print the result
print('entropy: %.3f' % Weak_entropy)

entropy: -0.000


# Calculating information gain for feature "temp"

In [73]:
# calculate the information gain
p_Strong=2/5
p_Weak=3/5
Average_entropy = (p_Strong * Strong_entropy) + (p_Weak * Weak_entropy)  
print('Average Entropy : %.3f' % Average_entropy)

Average Entropy : -0.000


In [74]:
Wind_Info_Gains = Sunny_entropy - Average_entropy
print('Information Gain : %.3f' % Wind_Info_Gains)

Information Gain : 0.971


In [75]:
Temp_Gain =  Temp_Info_Gains
Humidity_Gain = Humidity_Info_Gains
Wind_Gain = Wind_Info_Gains


print('Gain(S, Temp) : %.3f' % Temp_Gain)
print('Gain(S, Humidity) : %.3f' % Humidity_Gain)
print('Gain(S, Wind) : %.3f' % Wind_Gain)

Gain(S, Temp) : 0.020
Gain(S, Humidity) : 0.020
Gain(S, Wind) : 0.971


We can see that the label "wind" has the highest information gain of 0.971. Hence we can consider it as a root node

In [76]:
Rain_dataset.head()

temp humidity    wind play
3   Mild     High    Weak  Yes
4   Cool   Normal    Weak  Yes
5   Cool   Normal  Strong   No
9   Mild   Normal    Weak  Yes
13  Mild     High  Strong   No

# Decision Tree(ID3 Algorithm)

From the dataset, we can know that whenever there is weak wind, the label is yes and whenever the wind is strong, the label is no. Hence the tree can be drawn as:

![](final_diagram.png)